File: flux-tensor-midi/python/flux_tensor_midi/harmony/jaccard.py
"""
Jaccard similarity for FluxVector comparisons.

Jaccard index measures overlap between two sets.  Applied to
FluxVector space, it tells us how much two rooms' harmonic
content shares.
"""

from __future__ import annotations
import math
from flux_tensor_midi.core.flux import FluxVector


def jaccard_index(
    a: FluxVector,
    b: FluxVector,
    threshold: float = 0.01,
) -> float:
    """Jaccard similarity between two FluxVectors.

    Considers a channel "active" if its value is above threshold.

    Parameters
    ----------
    a, b : FluxVector
    threshold : float, default=0.01
        Minimum value to consider a channel active.

    Returns
    -------
    float
        Jaccard index (0 = disjoint, 1 = identical active sets).
    """
    active_a = {i for i, v in enumerate(a.values) if abs(v) > threshold}
    active_b = {i for i, v in enumerate(b.values) if abs(v) > threshold}

    intersection = active_a & active_b
    union = active_a | active_b

    if not union:
        return 1.0  # both silent = perfect agreement

    return len(intersection) / len(union)


def weighted_jaccard(
    a: FluxVector,
    b: FluxVector,
) -> float:
    """Weighted Jaccard using min/max of salience-weighted values.

    Treats each channel as having a magnitude; the index is
    sum(min(a_i, b_i)) / sum(max(a_i, b_i)).

    Returns 1.0 if both are all-zero.
    """
    a_vals = [v * s for v, s in zip(a.values, a.salience)]
    b_vals = [v * s for v, s in zip(b.values, b.salience)]

    numerator = sum(min(av, bv) for av, bv in zip(a_vals, b_vals))
    denominator = sum(max(av, bv) for av, bv in zip(a_vals, b_vals))

    if denominator == 0.0:
        return 1.0

    return numerator / denominator


def jaccard_distance(a: FluxVector, b: FluxVector, threshold: float = 0.01) -> float:
    """Jaccard distance = 1 - Jaccard index."""
    return 1.0 - jaccard_index(a, b, threshold)

File: flux-tensor-midi/rust/src/harmony/jaccard.rs
/// Jaccard similarity for flux channel sets.
///
/// Measures the overlap between two sets of active channels (or channel
/// properties) as a harmonic distance metric.

use crate::core::FluxVector;

/// Compute the Jaccard index between two sets of active indices.
///
/// `J(A, B) = |A ∩ B| / |A ∪ B|`
///
/// Returns 0.0 when both sets are empty (no active channels in either).
pub fn jaccard_active(a: &FluxVector, b: &FluxVector) -> f64 {
    let a_active: Vec<usize> = a
        .channels
        .iter()
        .enumerate()
        .filter(|(_, ch)| ch.intensity > 0)
        .map(|(i, _)| i)
        .collect();

    let b_active: Vec<usize> = b
        .channels
        .iter()
        .enumerate()
        .filter(|(_, ch)| ch.intensity > 0)
        .map(|(i, _)| i)
        .collect();

    if a_active.is_empty() && b_active.is_empty() {
        return 1.0; // two silences are "the same"
    }

    let intersection_count = a_active
        .iter()
        .filter(|i| b_active.contains(i))
        .count();

    let union_count = a_active.len() + b_active.len() - intersection_count;

    if union_count == 0 {
        return 0.0;
    }

    intersection_count as f64 / union_count as f64
}

/// Weighted Jaccard similarity that considers intensity values, not just
/// active/inactive status.
///
/// Uses the minimum-sum-over-maximum-sum formulation:
/// `J_w(A, B) = Σ min(A_i, B_i) / Σ max(A_i, B_i)`
/// where intensities are normalized to [0, 1].
pub fn weighted_jaccard(a: &FluxVector, b: &FluxVector) -> f64 {
    let mut sum_min = 0.0;
    let mut sum_max = 0.0;

    for (ch_a, ch_b) in a.channels.iter().zip(b.channels.iter()) {
        let na = ch_a.normalized();
        let nb = ch_b.normalized();
        sum_min += na.min(nb);
        sum_max += na.max(nb);
    }

    if sum_max == 0.0 {
        return 1.0;
    }

    sum_min / sum_max
}

/// Harmonic distance between two vectors as 1 - Jaccard (dissimilarity).
#[inline]
pub fn harmonic_distance(a: &FluxVector, b: &FluxVector) -> f64 {
    1.0 - jaccard_active(a, b)
}

/// Weighted harmonic distance.
#[inline]
pub fn weighted_harmonic_distance(a: &FluxVector, b: &FluxVector) -> f64 {
    1.0 - weighted_jaccard(a, b)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::FluxChannel;

    fn make_flux(active: &[usize]) -> FluxVector {
        let mut chs = [FluxChannel::new(0); 9];
        for &i in active {
            chs[i] = FluxChannel::new(100);
        }
        FluxVector::new(chs)
    }

    #[test]
    fn test_jaccard_identical() {
        let a = make_flux(&[0, 1, 2]);
        let b = make_flux(&[0, 1, 2]);
        assert!((jaccard_active(&a, &b) - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_jaccard_disjoint() {
        let a = make_flux(&[0, 1]);
        let b = make_flux(&[3, 4]);
        assert!((jaccard_active(&a, &b) - 0.0).abs() < 1e-10);
    }

    #[test]
    fn test_jaccard_partial() {
        let a = make_flux(&[0, 1, 2]);
        let b = make_flux(&[1, 2, 3]);
        // intersection = {1, 2} = 2, union = {0, 1, 2, 3} = 4
        assert!((jaccard_active(&a, &b) - 0.5).abs() < 1e-10);
    }

    #[test]
    fn test_jaccard_both_silent() {
        let a = make_flux(&[]);
        let b = make_flux(&[]);
        assert!((jaccard_active(&a, &b) - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_weighted_jaccard_full() {
        let a = FluxVector::uniform(127);
        let b = FluxVector::uniform(127);
        assert!((weighted_jaccard(&a, &b) - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_weighted_jaccard_half() {
        let a = FluxVector::uniform(127);
        let b = FluxVector::uniform(64);
        let wj = weighted_jaccard(&a, &b);
        // With uniform intensities, all pairs have the same ratio
        assert!(wj > 0.0 && wj < 1.0);
    }

    #[test]
    fn test_harmonic_distance() {
        let a = make_flux(&[0, 1]);
        let b = make_flux(&[0, 1]);
        assert!((harmonic_distance(&a, &b) - 0.0).abs() < 1e-10);
    }

    #[test]
    fn test_harmonic_distance_max() {
        let a = make_flux(&[0]);
        let b = make_flux(&[1]);
        assert!((harmonic_distance(&a, &b) - 1.0).abs() < 1e-10);
    }
}
